### ISSUE TITLE
Training Support for BetterTransformer

### ISSUE THREAD

**Issue by eugene-yang** [huggingface/optimum#971]
### System Info

```shell
Optimum main branch with pytorch=2.0 and cuda version 11.7.
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Running training with a model transformed with `BetterTransformer.transform` yields an error message on not supporting training (https://github.com/huggingface/optimum/blob/main/optimum/bettertransformer/models/base.py#L134). 

However, the documentation states that BetterTransformer already supports training. 
https://github.com/huggingface/optimum/blob/main/docs/source/bettertransformer/tutorials/convert.mdx#training-compatibility

### Expected behavior

Either the documentation should be updated or feature branch should be merged to main.

**Comment by fxmarty** [comment]
Hi @eugene-yang , which architecture are you trying with? Currently, only decoder models are supported for training. It may be worth upgrading optimum since we did a release today.

**Comment by fxmarty** [comment]
Yes, we should upgrade encoder models as well so that they are trainable. cc @nbroad1881

**Comment by Ananthzeke** [comment]
Still encoder based models are not compatible with training.  It throws an value error  
`Training is not supported for `BetterTransformer` integration.', ' Please use `model.eval()` before 
running the model.`

**Comment by metemadi** [comment]
hello and huge thank you to the amazing huggingface team as always for making everything so easy to use! any update on using BetterTransformer (and therefore memory efficient attention) for training? (for encoders and encoder/decoders?)

**Comment by canghongjian** [comment]
I have found that there is still no training support for BetterTransformer(Roberta).  The code seems like it supports decoder-based models, so your document confused me as it said:
![image](https://github.com/huggingface/optimum/assets/34473193/b55fb7d5-e5c9-4cca-ba29-43503734c6e0)
Is there any suggestions if I want to train an encoder-based model using FlashAttention ?

**Comment by fxmarty** [comment]
Hi, fixed in https://github.com/huggingface/optimum/pull/1225

However, for encoders, we use PyTorch SDPA with the passed attention mask in the training case. SDPA can currently not dispatch to flash attention when using an attention mask, so we either need to:
* ignore the attention mask altogether, trusting the user not to pass padded inputs (quite a constraint)
* use directy Hazy-flash implementation, that do support padding

**Comment by nbroad1881** [comment]
Is passing attention mask still an issue? I thought nested tensors were used so flash attention could still be used
